panel/auth.py

```diff
@@ -1069,6 +911,70 @@ class OAuthProvider(BasicAuthProvider):
     the global config.oauth_provider configuration.
     """
 
+    @property
+    def get_user(self):
+        return None
+
+    @property
+    def get_user_async(self):
+        async def get_user(handler):
+            user = handler.get_secure_cookie('user', max_age_days=config.oauth_expiry)
+            user = user.decode('utf-8') if user else None
+            if user and isinstance(handler, WebSocketHandler):
+                state._active_users[user] += 1
+            if not config.oauth_refresh_tokens or user is None:
+                return user
+
+            now_ts = dt.datetime.now(dt.timezone.utc).timestamp()
+            if user in state._oauth_user_overrides:
+                while not state._oauth_user_overrides[user]:
+                    await asyncio.sleep(0.1)
+                access_token = state._oauth_user_overrides[user]['access_token']
+            else:
+                access_cookie = handler.get_secure_cookie('access_token', max_age_days=config.oauth_expiry)
+                if not access_cookie:
+                    log.debug("No access token available, forcing user to reauthenticate.")
+                    return
+                access_token = state._decrypt_cookie(access_cookie)
+
+            try:
+                access_json = decode_token(access_token)
+            except (UnicodeDecodeError, ValueError):
+                # Token does not have content and therefore does not expire
+                log.debug("access_token is not a valid JWT token. Expiry cannot be determined.")
+                return user
+
+            if user in state._oauth_user_overrides:
+                refresh_token = state._oauth_user_overrides[user]['refresh_token']
+            else:
+                refresh_cookie = handler.get_secure_cookie('refresh_token', max_age_days=config.oauth_expiry)
+                if refresh_cookie:
+                    refresh_token = state._decrypt_cookie(refresh_cookie)
+                    self._schedule_refresh(access_json['exp'], user, refresh_token, handler.application, handler.request)
+                else:
+                    refresh_token = None
+
+            if access_json['exp'] > now_ts:
+                log.debug("Fully authenticated and access_token still valid.")
+                return user
+
+            if refresh_token:
+                try:
+                    refresh_json = decode_token(refresh_token)
+                    if refresh_json['exp'] < now_ts:
+                        refresh_token = None
+                except (ValueError, UnicodeDecodeError):
+                    pass
```

Comment:
Shouldn't `refresh_token` be set to None here if something goes wrong? With a warning

Reply:
No, I'll add a comment.  If the refresh_token is not a valid JWT token (which is what would trigger this error) then according to the spec it will not expire at all. I'll comment to that effect.